src/Orleans/Configuration/GlobalConfiguration.cs

```diff
@@ -670,13 +671,70 @@ internal override void Load(XmlElement root)
                         {
                             var providerConfig = new ProviderCategoryConfiguration();
                             providerConfig.Load(child);
-                            ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+
+                            if (ProviderConfigurations.ContainsKey(providerConfig.Name))
+                            {
+                                var existingProviderConfig = ProviderConfigurations[providerConfig.Name];
+                                existingProviderConfig.Merge(providerConfig);
+                            }
+                            else
+                            {
+                                ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+                            }
                         }
                         break;
                 }
             }
         }
 
+        /// <summary>
+        /// Registers given type of <typeparamref name="T"/> as bootstrap provider
+        /// </summary>
+        /// <typeparam name="T">Non-abstract type which implements <see cref="IBootstrapProvider"/> interface</typeparam>
+        /// <param name="properties">Properties that will be passed to bootstrap provider upon initialization</param>
+        public void RegisterBootstrapProvider<T>(IDictionary<string, string> properties = null)
```

Comment:
You know better as you have been digging in the code. It looks like one of the intentions is to signal there aren't duplicates. What's your opinion, would it be a worthwhile refactoring to provide some kind of a ``ConfigurationContainer``? With a quick look on the Internet, [this](http://www.dotnetcurry.com/showarticle.aspx?ID=973) looks just the kind I'm thinking currently, with associated reasoning. In addition to that, such a class could check parameters and verify their legal combinations in one place.